src/Giraffe/EndpointRouting.fs

```diff
@@ -220,68 +221,76 @@ let routef
     let boxedHandler (o : obj) =
         let t = o :?> 'T
         routeHandler t
-    TemplateEndpoint (HttpVerb.NotSpecified, template, mappings, boxedHandler)
+    TemplateEndpoint (HttpVerb.NotSpecified, template, mappings, boxedHandler, [])
 
 let subRoute
     (path      : string)
     (endpoints : Endpoint list) : Endpoint =
-    NestedEndpoint (path, endpoints)
+    NestedEndpoint (path, endpoints, [])
 
 let rec applyBefore
     (httpHandler  : HttpHandler)
     (endpoint     : Endpoint) =
     match endpoint with
-    | SimpleEndpoint(v, p, h)      -> SimpleEndpoint(v, p, httpHandler >=> h)
-    | TemplateEndpoint(v, p, m, h) -> TemplateEndpoint(v, p, m, fun (o: obj) -> httpHandler >=> h o)
-    | NestedEndpoint(t, lst)       -> NestedEndpoint(t, List.map (applyBefore httpHandler) lst)
+    | SimpleEndpoint(v, p, h, ml)      -> SimpleEndpoint(v, p, httpHandler >=> h, ml)
+    | TemplateEndpoint(v, p, m, h, ml) -> TemplateEndpoint(v, p, m, (fun (o: obj) -> httpHandler >=> h o), ml)
+    | NestedEndpoint(t, lst, ml)       -> NestedEndpoint(t, List.map (applyBefore httpHandler) lst, ml)
 
 let rec applyAfter
     (httpHandler  : HttpHandler)
     (endpoint     : Endpoint) =
     match endpoint with
-    | SimpleEndpoint(v, p, h)      -> SimpleEndpoint(v, p, h >=> httpHandler)
-    | TemplateEndpoint(v, p, m, h) -> TemplateEndpoint(v, p, m, fun (o: obj) -> h o >=> httpHandler)
-    | NestedEndpoint(t, lst)       -> NestedEndpoint(t, List.map (applyAfter httpHandler) lst)
+    | SimpleEndpoint(v, p, h, ml)      -> SimpleEndpoint(v, p, h >=> httpHandler, ml)
+    | TemplateEndpoint(v, p, m, h, ml) -> TemplateEndpoint(v, p, m, (fun (o: obj) -> h o >=> httpHandler), ml)
+    | NestedEndpoint(t, lst, ml)       -> NestedEndpoint(t, List.map (applyAfter httpHandler) lst, ml)
+
+let rec addMetadata
+    (metadata: obj)
+    (endpoint: Endpoint) =
+    match endpoint with
+    | SimpleEndpoint(v, p, h, ml)      -> SimpleEndpoint(v, p, h, metadata::ml)
+    | TemplateEndpoint(v, p, m, h, ml) -> TemplateEndpoint(v, p, m, h, metadata::ml)
+    | NestedEndpoint(t, lst, ml)       -> NestedEndpoint(t, lst, metadata::ml)
 
```

Comment:
This PR all looks great. With this `addMetadata` function, how are we intending it to be used when composing an app:
```fsharp
GET => route "/" (text "Hello World") |> addMetadata something
```
where `something` is an object that describes the whole `HttpHandler` side. Or would there be the ability to use it in the DSL at the front before the handler part? Not that it would be better, I'm just wondering if this PR dictates the usage yet.

Reply:
`GET >> addMetadata something => route "/test" (text "Hello World")` should also work (because both `GET` and `addMetdata something` are `Endpoint -> Endpoint`

Reply:
Ah, makes sense. I was sleepy and my signature foo was off 😆